SDXL Inpainting VAE Normalization

AlekseyKorshuk opened the pull request:
# What does this PR do?

Since the release of Playground V2.5 with "custom" VAE we should normalize and denormalize latents. 
This is already implemented in `StableDiffusionXLPipeline`: https://github.com/huggingface/diffusers/blob/687bc2772721af584d649129f8d2a28ca56a9ad8/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L1230-L1243


## Who can review?

cc: @sayakpaul @patil-suraj

sayakpaul commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py:
```diff
@@ -308,6 +308,25 @@ def retrieve_timesteps(
     return timesteps, num_inference_steps
 
 
+def requires_vae_latents_normalization(vae):
```
I think we can definitely do this in place without having to delegate it to a method. As that way the readability of the code stays linear and the reader doesn't have to refer to another method to see what's going on.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py:
```diff
@@ -308,6 +308,25 @@ def retrieve_timesteps(
+           hasattr(vae.config, "latents_std") and vae.config.latents_std is not None
+
+
+def normalize_vae_latents(latents, latents_mean, latents_std):
+    latents_mean = latents_mean.to(device=latents.device, dtype=latents.dtype)
+    latents_std = latents_std.to(device=latents.device, dtype=latents.dtype)
+    latents = (latents - latents_mean) / latents_std
+    return latents
```
Same as above.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py:
```diff
@@ -308,6 +308,25 @@ def retrieve_timesteps(
+    return latents
+
+
+def denormalize_vae_latents(latents, latents_mean, latents_std):
+    latents_mean = latents_mean.to(device=latents.device, dtype=latents.dtype)
+    latents_std = latents_std.to(device=latents.device, dtype=latents.dtype)
+    latents = latents * latents_std + latents_mean
+    return latents
```
Same as above.

sayakpaul reviewed (commented):
Do you have some results for us to see how the Playground v2.5 checkpoint plays out with inpainting. 

I am okay with that changes you're introducing given the comments are addressed. 

Cc: @patil-suraj as well.

AlekseyKorshuk commented:
@sayakpaul Thank for commenting! 

Don't you think that the code should be DRY (Don't Repeat Yourself)? Overall these 3 functions should be used in any place of SDXL pipeline with this VAE. Maybe it is reasonable to move it to "utils"/"common" or in base SDXL pipeline.
Just my thoughts, otherwise happy to put inplace without delegating the function.

Don't have checkpoint yet. I trained one but figured out this normalisation too late...

sayakpaul commented:
> Don't you think that the code should be DRY (Don't Repeat Yourself)? Overall these 3 functions should be used in any place of SDXL pipeline with this VAE. Maybe it is reasonable to move it to "utils"/"common" or in base SDXL pipeline.

Please refer to https://huggingface.co/docs/diffusers/en/conceptual/philosophy. It's okay to trade away DRY in the interest of readability as our pipelines are also read for educational purposes.

AlekseyKorshuk commented:
Addressed all the comments. Please let me know if I can be helpful. 

Btw while you are replying, if it is possible by any chance to share training args used for this PR to train [inpainting model](https://huggingface.co/diffusers/stable-diffusion-xl-1.0-inpainting-0.1):
* https://github.com/huggingface/diffusers/pull/6592#issuecomment-1965476311

This would be very helpful, thank you!

sayakpaul commented:
> Btw while you are replying, if it is possible by any chance to share training args used for this PR to train [inpainting model](https://huggingface.co/diffusers/stable-diffusion-xl-1.0-inpainting-0.1):

Will defer that to the training ninja @patil-suraj here :)

yiyixuxu commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py:
```diff
@@ -939,6 +939,17 @@ def _encode_vae_image(self, image: torch.Tensor, generator: torch.Generator):
         else:
             image_latents = retrieve_latents(self.vae.encode(image), generator=generator)
 
+        has_latents_mean = hasattr(self.vae.config, "latents_mean") and self.vae.config.latents_mean is not None
```
why do we need it here?

sayakpaul replied:
I think this would be needed because here we are encoding the input image. This is what we do in the SDXL DreamBooth LoRA training script.

sayakpaul replied:
But I agree that #7132 tackles it from a comprehensive perspective.

yiyixuxu replied:
ok I re-opene it and I'm happy to be proven wrong
once we merge #7132, you can update the branch, and you can then test it out and show some results with and without the normalization.

sayakpaul replied:
Works for me!

yiyixuxu commented:
just saw this PR https://github.com/huggingface/diffusers/pull/7132
I think it is correctly done there I'm closing this PR for now. Feel free to leave any questions and we will continue the discussion here

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_7225). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sayakpaul commented:
@yiyixuxu could you give this another look?

sayakpaul commented:
@yiyixuxu a gentle ping.
